<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>A Post-clustering Differential Expression (DE) Method For Solving Double Dipping • ClusterDE</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="A Post-clustering Differential Expression (DE) Method For Solving Double Dipping">
<meta name="description" content="ClusterDE is a post-clustering DE method for controlling the false discovery rate (FDR) of identified DE genes regardless of clustering quality. The core idea of ClusterDE is to generate real-data-based synthetic null data with only one cluster, as a counterfactual in contrast to the real data, for evaluating the whole procedure of clustering followed by a DE test.">
<meta property="og:description" content="ClusterDE is a post-clustering DE method for controlling the false discovery rate (FDR) of identified DE genes regardless of clustering quality. The core idea of ClusterDE is to generate real-data-based synthetic null data with only one cluster, as a counterfactual in contrast to the real data, for evaluating the whole procedure of clustering followed by a DE test.">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">ClusterDE</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="articles/ClusterDE-cellline.html">Perform ClusterDE on a cell line dataset</a></li>
    <li><a class="dropdown-item" href="articles/ClusterDE-onedomain.html">Perform ClusterDE on a single-domain spatial dataset</a></li>
    <li><a class="dropdown-item" href="articles/ClusterDE-PBMC.html">Perform ClusterDE on a PBMC dataset</a></li>
    <li><a class="dropdown-item" href="articles/ClusterDE-twodomains.html">Perform ClusterDE on a two-domain spatial dataset</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/SONGDONGYUAN1994/ClusterDE/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="clusterde">ClusterDE<a class="anchor" aria-label="anchor" href="#clusterde"></a>
</h1></div>
<hr>
<p>The R package <strong>ClusterDE</strong> is a post-clustering DE method for controlling the false discovery rate (FDR) of identified between cell-type DE genes regardless of clustering quality. The core idea of ClusterDE is to generate real-data-based synthetic null data with only one cell type, as contrast to the real data, for evaluating the whole procedure of clustering followed by a DE test. <span style="color:blue"> <strong>Detailed tutorials that illustrate various functionalities of ClusterDE are available at this <a href="https://songdongyuan1994.github.io/ClusterDE/index.html">website</a></strong></span>. The following illustration figure summarizes the usage of ClusterDE:</p>
<p><img src="reference/figures/ClusterDE_illustration.png" width="600"></p>
<p>Instead of a new pipeline, ClusterDE actually works as an add-on to popular pipelines such as Seurat. To find out more details about <strong>ClusterDE</strong>, you can check out our manuscript on <a href="https://www.biorxiv.org/content/10.1101/2023.07.21.550107v1" class="external-link">bioRxiv</a>.</p>
</div>
<div class="section level1">
<h1 id="changelog">Changelog<a class="anchor" aria-label="anchor" href="#changelog"></a>
</h1>
<ul><li>2024-06-29 Important changes<ul><li>Add functions for spatial clustering</li></ul>
</li></ul>
<p><strong>The motivation and application of ClusterDE</strong>: In Seurat function <code>findMarkers</code>, the authors pointed out: <em>“p-values should be interpreted cautiously, as the genes used for clustering are the same genes tested for differential expression.”</em> This is the “double-dipping” issue. If your clustering results are inaccurate and since the clustering has used your expression data already, the discovered DE genes may not represent the discrete cell type separation, but other variation in your data (e.g., cell cycle, total UMI, or other variation you are not clear. These are still biological variation but do not define discrete status).</p>
<p>ClusterDE aims at correcting the double-dipping issue for comparing two dubious clusters, which you are not sure if they are two discrete cell types or just an artifact of your clustering algorithm based on conventional DE analysis. ClusterDE controls the false discoveries in DE and prioritizes the true cell type markers.</p>
<p>Note: current version is focusing on one vs one comparison.</p>
</div>
<div class="section level1">
<h1 id="installation">Installation<a name="installation-"></a>
<a class="anchor" aria-label="anchor" href="#installation"></a>
</h1>
<p>To install the development version from GitHub, please run:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://devtools.r-lib.org/" class="external-link">"devtools"</a></span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"SONGDONGYUAN1994/ClusterDE"</span><span class="op">)</span></span></code></pre></div>
<p>Please note that ClusterDE is actually a wrapper of <strong>scDesign3</strong>. Therefore, you can also directly use scDesign3 to " design" your own synthetic null data. To better understand scDesign3, you can check out our manuscript on Nature Biotechnology:</p>
<p><a href="https://www.nature.com/articles/s41587-023-01772-1" class="external-link">Song, D., Wang, Q., Yan, G. et al. scDesign3 generates realistic in silico data for multimodal single-cell and spatial omics. <em>Nat Biotechnol</em> (2023).</a></p>
</div>
<div class="section level1">
<h1 id="quick-start">Quick Start<a name="quick-start"></a>
<a class="anchor" aria-label="anchor" href="#quick-start"></a>
</h1>
<p>A basic way to identify differentially expressed genes (DEGs) between two clusters is to use <code><a href="reference/findMarkers.html">ClusterDE::findMarkers()</a></code> on a Seurat object that already contains the count matrix and clustering results.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># obj is a preprocessed Seurat object with clustering completed</span></span>
<span><span class="va">markers</span> <span class="op">&lt;-</span> <span class="fu">ClusterDE</span><span class="fu">::</span><span class="fu"><a href="reference/findMarkers.html">findMarkers</a></span><span class="op">(</span><span class="va">obj</span>, ident.1 <span class="op">=</span> <span class="fl">0</span>, ident.2 <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p>This generates a list of genes ranked by contrast score (cs), where higher scores indicate stronger differential expression.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">markers</span><span class="op">)</span></span>
<span><span class="co"># # A tibble: 6 × 3</span></span>
<span><span class="co">#   gene        cs record</span></span>
<span><span class="co">#   &lt;chr&gt;    &lt;dbl&gt;  &lt;dbl&gt;</span></span>
<span><span class="co"># 1 CDKN1C    39.9      1</span></span>
<span><span class="co"># 2 CD79B     38.6      1</span></span>
<span><span class="co"># 3 FCGR3A    33.8      1</span></span>
<span><span class="co"># 4 CKB       32.3      1</span></span>
<span><span class="co"># 5 SIGLEC10  24.4      1</span></span>
<span><span class="co"># 6 MTSS1     23.8      1</span></span></code></pre></div>
<ul>
<li>gene: gene symbol</li>
<li>cs: contrast score</li>
<li>record: indicator of reproducibility across null replicates</li>
</ul>
</div>
<div class="section level1">
<h1 id="other-usages">Other Usages<a class="anchor" aria-label="anchor" href="#other-usages"></a>
</h1>
<div class="section level2">
<h2 id="pca-approximation">PCA approximation<a class="anchor" aria-label="anchor" href="#pca-approximation"></a>
</h2>
<p>For faster and more stable DEG detection when using multiple null-data replicates, we recommend using the <strong>PCA-based approximation</strong> for generating synthetic null data. Enable this by setting the <code>flavour</code> argument in <code><a href="reference/findMarkers.html">findMarkers()</a></code> to <code>"pca"</code> (default is <code>"classic"</code>). Increase the number of threads (<code>nCores</code>) to speed up parallel calculation.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">markers</span> <span class="op">&lt;-</span> <span class="fu">ClusterDE</span><span class="fu">::</span><span class="fu"><a href="reference/findMarkers.html">findMarkers</a></span><span class="op">(</span><span class="va">obj</span>, ident.1 <span class="op">=</span> <span class="fl">0</span>, ident.2 <span class="op">=</span> <span class="fl">1</span>, flavour <span class="op">=</span> <span class="st">"pca"</span>, nCores <span class="op">=</span> <span class="fl">8</span>, nRep <span class="op">=</span> <span class="fl">40</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="customize-null-data-generation">Customize null data generation<a class="anchor" aria-label="anchor" href="#customize-null-data-generation"></a>
</h2>
<p>Beyond the default <code><a href="reference/findMarkers.html">findMarkers()</a></code> parameters, ClusterDE also allows users to customize the analysis pipeline, particularly the generation and inspection of synthetic null data.</p>
<p>The example below shows how to manually generate null data from a gene-by-cell matrix containing the two clusters of interest. For typical UMI count data, we recommend using the Negative Binomial model (<code>nb</code>). The synthetic null data generation is relatively time-consuming; you may use the fast version (<code>fastVersion = TRUE</code>).</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">exampleCounts</span><span class="op">)</span></span>
<span><span class="va">nullData</span> <span class="op">&lt;-</span> <span class="fu">ClusterDE</span><span class="fu">::</span><span class="fu"><a href="reference/constructNull.html">constructNull</a></span><span class="op">(</span></span>
<span>  mat <span class="op">=</span> <span class="va">exampleCounts</span>,</span>
<span>  family <span class="op">=</span> <span class="st">"nb"</span>,</span>
<span>  formula <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  extraInfo <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  nCores <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  parallelization <span class="op">=</span> <span class="st">"mcmapply"</span>,</span>
<span>  fastVersion <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  corrCut <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>  BPPARAM <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>The parameters of <code><a href="reference/constructNull.html">constructNull()</a></code> are:</p>
<ul>
<li>
<code>mat</code>: The input gene by cell matrix. It can be a sparse matrix.</li>
<li>
<code>family</code>: A string of the distribution you want to use when fitting the model. Must be one of ‘poisson’, ‘nb’, ‘zip’, ‘zinb’ or ‘gaussian’.</li>
<li>
<code>formula</code>: A string of the mu parameter formula. It defines the relationship between gene expression in synthetic null data and the extra covariates. Default is NULL (cell type case). For example, if your input data is a spatial data with X, Y coordinates, the formula can be ‘s(X, Y, bs = ’gp’, k = 4)’.</li>
<li>
<code>extraInfo</code>: A data frame of the extra covariates used in . For example, the 2D spatial coordinates. Default is NULL.</li>
<li>
<code>nCores</code>: An integer. The number of cores to use. Increasing the cores will greatly speed up the computation.</li>
<li>
<code>parallelization</code>: A string indicating the specific parallelization function to use. Must be one of ‘mcmapply’, ‘bpmapply’, or ‘pbmcmapply’, which corresponds to the parallelization function in the package ‘parallel’, ‘BiocParallel’, and ‘pbmcapply’ respectively. The default value is ‘pbmcmapply’.</li>
<li>
<code>fastVersion</code>: A logic value. If TRUE, the fast approximation is used.</li>
<li>
<code>corrCut</code>: A numeric value. The cutoff for non-zero proportions in genes used in modelling correlation. The reason is that lowly expressed genes are hard to calculate correlation.</li>
<li>
<code>BPPARAM</code>: A MulticoreParam object or NULL. When the parameter parallelization = ‘mcmapply’ or ‘pbmcmapply’, this parameter must be NULL. When the parameter parallelization = ‘bpmapply’, this parameter must be one of the MulticoreParam object offered by the package ‘BiocParallel’. The default value is NULL.</li>
<li>
<code>approximation</code>: TBA</li>
<li>
<code>usePca</code>: TBA</li>
<li>
<code>nPcs</code>: TBA</li>
</ul>
<p>The output of <code><a href="reference/constructNull.html">constructNull()</a></code> is the new gene by cell matrix in the same format as your input.</p>
<p>The following figure briefly describes how ClusterDE generates the synthetic null data:</p>
<p><img src="reference/figures/ClusterDE_supp_null_generation.png" width="800"></p>
<p>After obtaining the synthetic null data, use <code><a href="reference/calcNullPval.html">calcNullPval()</a></code> to perform standard preprocessing and clustering pipeline to get DE p-values (<code>nullPvalues</code>). Finally, we compare the p-values from the null and p-values from the target data (real data) by <code><a href="reference/callDE.html">callDE()</a></code>. For illustration, here we use the Uniform random numbers as the p-values.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span><span class="va">targetPvalues</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/FindMarkers.html" class="external-link">FindMarkers</a></span><span class="op">(</span><span class="va">obj</span>, ident.1 <span class="op">=</span> <span class="fl">0</span>, ident.2 <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">nullPvalues</span> <span class="op">&lt;-</span> <span class="fu">ClusterDE</span><span class="fu">::</span><span class="fu"><a href="reference/calcNullPval.html">calcNullPval</a></span><span class="op">(</span><span class="va">nullData</span><span class="op">)</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">ClusterDE</span><span class="fu">::</span><span class="fu"><a href="reference/callDE.html">callDE</a></span><span class="op">(</span></span>
<span>  targetScores <span class="op">=</span> <span class="va">targetPvalues</span>,</span>
<span>  nullScores <span class="op">=</span> <span class="va">nullPvalues</span>,</span>
<span>  nlogTrans <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  FDR <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  correct <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>The parameters of <code>callDE</code> are:</p>
<ul>
<li>
<code>targetScores</code>: A named numeric vector of the DE scores from the target data, e.g., the p-values between two clusters from the real data.</li>
<li>
<code>nullScores</code>: A named numeric vector of the DE scores from the synthetic null data, e.g., the p-values between two clusters from the null data.</li>
<li>
<code>nlogTrans</code>: A logical value. If the input scores are p-values, take the <code>-log10</code> transformation since Clipper require larger scores represent more significant DE. Default is TRUE.</li>
<li>
<code>FDR</code>: A numeric value of the target False Discovery Rate (FDR). Default is 0.05.</li>
<li>
<code>correct</code>: A logical value. If TRUE, perform the correction to make the distribution of contrast scores approximately symmetric. Default is FALSE.</li>
</ul>
<p>The output of <code><a href="reference/callDE.html">callDE()</a></code> is a list of DE genes, same as the output of <code><a href="reference/findMarkers.html">findMarkers()</a></code>.</p>
</div>
</div>
<div class="section level1">
<h1 id="tutorials">Tutorials<a name="tutorials"></a>
<a class="anchor" aria-label="anchor" href="#tutorials"></a>
</h1>
<p>For all detailed tutorials, please check the <a href="https://songdongyuan1994.github.io/ClusterDE/index.html">website</a>. The tutorials will demonstrate the applications of <strong>ClusterDE</strong> in two cases: a cell line dataset (no cell type exists) and a PBMC dataset.</p>
<ul>
<li><a href="https://songdongyuan1994.github.io/ClusterDE/articles/ClusterDE-cellline.html">Perform ClusterDE on a cell line dataset</a></li>
<li><a href="https://songdongyuan1994.github.io/ClusterDE/articles/ClusterDE-PBMC.html">Perform ClusterDE on a PBMC dataset</a></li>
<li><a href="https://songdongyuan1994.github.io/ClusterDE/articles/ClusterDE-onedomain.html">Perform ClusterDE on a single-domain spatial transcriptomics dataset</a></li>
<li><a href="https://songdongyuan1994.github.io/ClusterDE/articles/ClusterDE-twodomains.html">Perform ClusterDE on a two-domains spatial transcriptomics dataset</a></li>
</ul>
</div>
<div class="section level1">
<h1 id="contact">Contact<a name="contact"></a>
<a class="anchor" aria-label="anchor" href="#contact"></a>
</h1>
<p>Any questions or suggestions on <code>ClusterDE</code> are welcomed! Please report it on <a href="https://github.com/SONGDONGYUAN1994/ClusterDE/issues" class="external-link">issues</a>, or contact Dongyuan Song (<a href="mailto:dongyuansong@ucla.edu" class="email">dongyuansong@ucla.edu</a>).</p>
</div>
<div class="section level1">
<h1 id="related-papers">Related Papers<a name="related-manuscripts"></a>
<a class="anchor" aria-label="anchor" href="#related-papers"></a>
</h1>
<ul>
<li>
<strong>scDesign3</strong>: <a href="https://www.nature.com/articles/s41587-023-01772-1" class="external-link">Song, D., Wang, Q., Yan, G. et al. scDesign3 generates realistic in silico data for multimodal single-cell and spatial omics. <em>Nat Biotechnol</em> (2023).</a>
</li>
<li>
<strong>Clipper</strong>: <a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-021-02506-9" class="external-link">Ge, X., Chen, Y.E., Song, D. et al. Clipper: p-value-free FDR control on high-throughput data from two conditions. <em>Genome Biology</em> 22, 288 (2021).</a>
</li>
</ul>
</div>
<div class="section level1">
<h1 id="other-methods-for-double-dipping-problem">Other methods for double dipping problem<a class="anchor" aria-label="anchor" href="#other-methods-for-double-dipping-problem"></a>
</h1>
<ul>
<li>
<strong>TN test</strong>: <a href="https://www.sciencedirect.com/science/article/pii/S2405471219302698" class="external-link">Zhang, J.M., Kamath, G.M., David, N.T. Valid post-clustering differential analysis for single-cell rna-seq. <em>Cell Systems</em>, 2019</a>
</li>
<li>
<strong>count split</strong>: <a href="https://academic.oup.com/biostatistics/advance-article/doi/10.1093/biostatistics/kxac047/6893953?login=true" class="external-link">Neufeld, A., Gao, L.L., Popp, J., Battle, A., Witten, D. Inference after latent variable estimation for single-cell RNA sequencing data. <em>Biostatistics</em>, 2022</a>
</li>
</ul>
</div>

  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/SONGDONGYUAN1994/ClusterDE/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/SONGDONGYUAN1994/ClusterDE/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing ClusterDE</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Dongyuan Song <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0003-1114-1215" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a>  </li>
<li>Kexin Li <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0002-4955-3832" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a>  </li>
<li>Siqi Chen <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0002-9498-3823" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a>  </li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Dongyuan Song, Kexin Li, Siqi Chen.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
