---
title: "Perform ClusterDE on a two-domain spatial dataset"
author: 
  - name: Siqi Chen
    affiliation:
    - Computer Science, Central South University
    email: siqichen4477@gmail.com
  - name: Dongyuan Song
    affiliation:
    - Department of Genetics & Genome Sciences, UConn Health 
    - Bioinformatics IDP, University of California, Los Angeles
    email: dongyuansong@ucla.edu
output: 
  BiocStyle::html_document:
    self_contained: yes
    toc: true
    toc_float: true
    toc_depth: 2
    code_folding: show
date: "`r doc_date()`"
package: "`r pkg_ver('ClusterDE')`"
vignette: >
  %\VignetteIndexEntry{ClusterDE-twodomains_vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}  
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
tools::R_user_dir("ClusterDE", which="cache")
```

# Download data

We selected adjacent layer 6 and WM of 151673 slice from the LIBD Human Dorsolateral Prefrontal Cortex (DLPFC) dataset, which is downloaded in the spatialLIBD R package. We removed the genes expressed in fewer than 20% spots to save computational time here.
```{r}
# # Download the spot-level data
# spe <- spatialLIBD::fetch_data(type = "spe")
#
# # Select the Layer6 and WM domains in the slice "151673"
# sub_151673 <- spe[, spe$sample_id == "151673"]
# index <- sub_151673$spatialLIBD == "L6" | sub_151673$spatialLIBD == "WM"
# index[which(is.na(index))] <- "NAN"
# sub_151673 <- sub_151673[, index == "TRUE"]
# print(sub_151673)
#
# # Delete the genes that express rate less than 20%
# data <- sub_151673@assays@data$counts
# zero_expre_rate <- apply(data, 1, function(x) {
#   zero_true <- x == 0
#   zero_num <- length(which(zero_true == TRUE)) / dim(data)[2]
#   return(zero_num)
# })
# zero_expre_gene_idx <- which(zero_expre_rate < 0.8)
# sub_151673 <- sub_151673[zero_expre_gene_idx,]
# cat(paste0("The size of data: ", dim(sub_151673)[1], "*", dim(sub_151673)[2], sep = ""))
#
# # Construct the SingleCellExperiment object
# dlpfc_twodomain <- SingleCellExperiment::SingleCellExperiment(list(counts = sub_151673@assays@data$counts))
# # Add colData information of singlecellexperiment
# dlpfc_twodomain$spatial1 <- sub_151673@int_colData@listData$spatialCoords[, 2]
# dlpfc_twodomain$spatial2 <- sub_151673@int_colData@listData$spatialCoords[, 1]
# dlpfc_twodomain$cell_type <- sub_151673@colData$spatialLIBD
# SingleCellExperiment::logcounts(dlpfc_twodomain) <- log1p(SingleCellExperiment::counts(dlpfc_twodomain))
# dlpfc_twodomain <- Seurat::as.Seurat(dlpfc_twodomain)
data(dlpfc_twodomain, package = "ClusterDE")
```

Visualize the real data with two domains (L6 and WM).
```{r}
# Visualize the real spatial domains
domains <- data.frame(Xaxis = dlpfc_twodomain$spatial1, Yaxis = dlpfc_twodomain$spatial2, Domains = dlpfc_twodomain$cell_type)
ggplot2::ggplot(domains, ggplot2::aes(x = Xaxis, y = Yaxis, col = Domains)) +
  ggplot2::geom_point(size = 1.0) +
  ggplot2::coord_equal() +
  ggplot2::ggtitle("Manual annotation \n (The layer6 and WM in the slice 151673)") +
  ggplot2::theme(
    plot.title = ggplot2::element_text(size = 10, hjust = 0.5),
    panel.grid = ggplot2::element_blank(),
    panel.background = ggplot2::element_rect(fill = "gray90"),
    panel.border = ggplot2::element_rect(color = "black", fill = NA, linewidth = 0.6),
    axis.title.x = ggplot2::element_blank(),
    axis.title.y = ggplot2::element_blank(),
    axis.ticks.x = ggplot2::element_blank(),
    axis.ticks.y = ggplot2::element_blank(),
    axis.text.x = ggplot2::element_blank(),
    axis.text.y = ggplot2::element_blank()
  ) +
  ggplot2::scale_color_manual(values = c("#5791cc", "#403f92"))
```

# Run the BayesSpace + Seurat pipeline

Firstly, we employed the BayesSpace for spatial clustering. Please note that ClusterDE is designed for 1 vs 1 comparison; therefore, we obtain two spatial clusters for illustration purpose.
```{r}
# Construct the input of BayesSpace based on real dataset
# The input of BayesSpace is sce object
dlpfc_cluster <- SingleCellExperiment::SingleCellExperiment(
  list(counts = Seurat::GetAssayData(dlpfc_twodomain, layer = "counts"))
)
# Add colData information of singlecellexperiment
dlpfc_cluster$row <- dlpfc_twodomain$spatial1
dlpfc_cluster$col <- dlpfc_twodomain$spatial2

# Log-normalize the count data
set.seed(123)
dlpfc_cluster <- BayesSpace::spatialPreprocess(dlpfc_cluster, platform = "ST", n.PCs = 7, log.normalize = T)
# Clustering with BayesSpace
dlpfc_cluster <- BayesSpace::spatialCluster(
  dlpfc_cluster,
  q = 2,
  platform = "ST",
  d = 7,
  init.method = "mclust",
  model = "t",
  gamma = 2,
  nrep = 1000,
  burn.in = 100,
  save.chain = T
)
```

Visualize the spatial clustering results based on the real data.
```{r}
# Visualize the spatial cluster
clusters <- data.frame(Xaxis = dlpfc_cluster$row, Yaxis = dlpfc_cluster$col, Clusters = as.character(dlpfc_cluster$spatial.cluster))
ggplot2::ggplot(clusters, ggplot2::aes(x = Xaxis, y = Yaxis, col = Clusters)) +
  ggplot2::geom_point(size = 1.0) +
  ggplot2::coord_equal() +
  ggplot2::ggtitle("Real data \n (spatial clusters detected by BayesSpace)") +
  ggplot2::theme(
    plot.title = ggplot2::element_text(size = 10, hjust = 0.5),
    panel.grid = ggplot2::element_blank(),
    panel.background = ggplot2::element_rect(fill = "gray90"),
    panel.border = ggplot2::element_rect(color = "black", fill = NA, size = 0.6),
    axis.title.x = ggplot2::element_blank(),
    axis.title.y = ggplot2::element_blank(),
    axis.ticks.x = ggplot2::element_blank(),
    axis.ticks.y = ggplot2::element_blank(),
    axis.text.x = ggplot2::element_blank(),
    axis.text.y = ggplot2::element_blank()
  ) +
  ggplot2::scale_color_manual(values = c("#e87d72", "#54bcc2"))
```

Then, we used the common DE method (Wilcoxon Rank Sum Test) to identify domain marker genes between the two spatial clusters.
```{r}
# Identify domain marker genes in the real dataset based on the BayesSpace clustering result, follow Seurat tutorial
dlpfc_twodomain <- Seurat::as.Seurat(dlpfc_cluster)
Seurat::Idents(dlpfc_twodomain) <- "spatial.cluster"
original_markers <- Seurat::FindMarkers(
  object = dlpfc_twodomain,
  ident.1 = 1,
  ident.2 = 2,
  test.use = "wilcox",
  logfc.threshold = 0,
  min.pct = 0,
  min.cells.feature = 1,
  min.cells.group = 1
)
head(original_markers)
```

# Find DEGs using ClusterDE

We can use `findMarkers()` from ClusterDE with `spatial` specified as name of X and Y coordinates in the data meta columns. Here in the example, coordinates are stored in `row` and `col`.
```{r}
res <- ClusterDE::findMarkers(dlpfc_twodomain, ident.1 = 1, ident.2 = 2, spatial = c("row", "col"))
```

# Visualize top marker genes

We visualize the top DE genes from ClusterDE. As expected, the top genes detected by ClusterDE exhibit clear spatial expression patterns.
```{r}
top_genes_clusterde <- res$gene[1:6]
expr <- Seurat::GetAssayData(dlpfc_twodomain, layer = "data")[top_genes_clusterde, , drop = F]
x_coord <- dlpfc_twodomain@meta.data$row
y_coord <- dlpfc_twodomain@meta.data$col

plots <- lapply(top_genes_clusterde, function(gene) {
  df <- data.frame(
    X = x_coord,
    Y = y_coord,
    Expr = as.numeric(expr[gene,])
  )
  ggplot2::ggplot(df, ggplot2::aes(x = X, y = Y, color = Expr)) +
    ggplot2::geom_point(size = 0.5) +
    ggplot2::scale_colour_gradientn(colors = viridis::viridis_pal(option = "magma")(10)) +
    ggplot2::coord_fixed() +
    ggplot2::ggtitle(gene) +
    ggplot2::theme_void() +
    ggplot2::theme(legend.position = "bottom")
})

patchwork::wrap_plots(plots, nrow = 1)
```

# Session information

```{r}
sessionInfo()
```


