---
title: "Perform ClusterDE on a PBMC dataset"
author: 
  - name: Dongyuan Song
    affiliation:
    - Department of Genetics & Genome Sciences, UConn Health 
    - Bioinformatics IDP, University of California, Los Angeles
    email: dongyuansong@ucla.edu
output: 
  BiocStyle::html_document:
    self_contained: yes
    toc: true
    toc_float: true
    toc_depth: 2
    code_folding: show
date: "`r doc_date()`"
package: "`r pkg_ver('ClusterDE')`"
vignette: >
  %\VignetteIndexEntry{ClusterDE-PBMC_vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}  
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
tools::R_user_dir("ClusterDE", which = "cache")
```

```{r setup, message=FALSE, warning=FALSE, results='hide'}
#tools::R_user_dir("ClusterDE", which="cache")
library(ClusterDE)
library(Seurat)
library(SingleCellExperiment)
library(ggplot2)
library(patchwork)
#library(SeuratData)
library(Matrix)
```

# Download data
The PBMC datasets are originally from `SeuratData`. We use one of them (10x Chromium (v3) from PBMC1 replicate). We filtered out some lowly epxressed genes to save computational time here.
```{r}
# InstallData("pbmcsca")
# data("pbmcsca")
# pbmc <- pbmcsca[, pbmcsca@meta.data$Method=="10x Chromium (v3)" & pbmcsca@meta.data$Experiment == "pbmc1"]
# 
# pbmc <- pbmc[Matrix::rowSums(pbmc@assays$RNA@counts) > 100, ]

pbmc <- readRDS(url("https://figshare.com/ndownloader/files/41486283"))
```

# Run the regular Seurat pipeline

We perform the default Seurat clustering. Note that in real data analysis, the cell type label is usually unknown.

```{r, fig.width=12, fig.height=6}
RNGkind("L'Ecuyer-CMRG")
set.seed(123)
pbmc <- UpdateSeuratObject(pbmc) ## Avoid errors
pbmc <- NormalizeData(pbmc)
pbmc <- FindVariableFeatures(object = pbmc)
pbmc <- ScaleData(object = pbmc)
pbmc <- RunPCA(object = pbmc)
pbmc <- FindNeighbors(object = pbmc)
pbmc <- FindClusters(object = pbmc, resolution = 0.3)
pbmc <- RunUMAP(object = pbmc, dims = 1:10)
p1 <- DimPlot(object = pbmc, reduction = "umap", label = TRUE) +
  ggtitle("Clustering result") +
  NoLegend()
p2 <- DimPlot(object = pbmc, reduction = "umap", group.by = "CellType", label = TRUE) + NoLegend()
p1 + p2
```

In this vignette, we are interested in cluster 2 vs 8, which approximately represent CD14+/CD16+ monocytes. Please note that ClusterDE is designed for 1 vs 1 comparison. Therefore, users may (1) choose the two interested clusters manually based on their knowledge or (2) use the two locally closest clusters from computation (e.g., `BuildClusterTree` in `Seurat`).
```{r}
pbmc <- BuildClusterTree(pbmc)
PlotClusterTree(pbmc)
```

We perform the DE test between cluster 2 and 8. We subset the cluster 2 and 8 (`pbmc_sub`).
```{r}
pbmc_sub <- subset(x = pbmc, idents = c(2, 8))
original_markers <- FindMarkers(
  pbmc_sub,
  ident.1 = 2,
  ident.2 = 8,
  min.pct = 0,
  logfc.threshold = 0
)
```
**Note**: since we need the same number of genes being tested in real data and null data, we manually remove the default filtering in Seurat, which will exclude some genes from being tested. We believe that this filtering should be performed after you test every gene.

# Find DEGs using ClusterDE

We can use `findMarkers()` to perform null-calibrated post-clustering differential expression. The result table is sorted by contrast scores.
```{r}
res <- ClusterDE::findMarkers(pbmc_sub, ident.1 = 2, ident.2 = 8)
```

To compare the result from the naive Seurat pipeline and ClusterDE, we first visualize the top 6 DE genes from Seurat. Genes *LST1* and *RPS19* are both highly expressed in two clusters. In addition, *RPS19* is reported as a stable housekeeping genes in several studies. Note that it does not mean the expression levels of *LST1* and *RPS19* are the same between the two cell types. It means that they are not good cell type markers. Philosophically speaking, it means that **conditional on the two clusters are obtained by clustering algorithm, LST1 and RPS19 are less likely to be the cell type markers between the two cell types**.
```{r, fig.width=10, fig.height=6}
FeaturePlot(pbmc[, pbmc$seurat_clusters %in% c(2, 8)], features = c(rownames(original_markers)[1:6]), ncol = 3)
```
In contrast, the genes from ClusterDE do not have *LST1* and *RPS19* anymore.
```{r, fig.width=10, fig.height=6}
FeaturePlot(pbmc[, pbmc$seurat_clusters %in% c(2, 8)], features = c(res$gene[1:6]), ncol = 3)
```

# Session information
```{r}
sessionInfo()
```
