---
title: "Perform ClusterDE on a cell line dataset"
author: 
  - name: Dongyuan Song
    affiliation:
    - Department of Genetics & Genome Sciences, UConn Health 
    - Bioinformatics IDP, University of California, Los Angeles
    email: dongyuansong@ucla.edu
output: 
  BiocStyle::html_document:
    self_contained: yes
    toc: true
    toc_float: true
    toc_depth: 2
    code_folding: show
date: "`r doc_date()`"
package: "`r pkg_ver('ClusterDE')`"
vignette: >
  %\VignetteIndexEntry{ClusterDE-cellline_vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}  
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
tools::R_user_dir("ClusterDE", which = "cache")
```

# Download data

We download the cell line data set H2228. The original data is from [Tian et al., Nature Methods 2019](https://github.com/LuyiTian/sc_mixology) as the gold standard for benchmarking the accuracy of clustering. Since the data is from the pure cell line, it should not have cell types, and, of course, between cell type DE genes. **Note**: it does not mean that there are no variations within one cell line; of course there are, e.g., cell cycle, total UMI, etc. However, these variations do not represent **discrete cell groups**, and essentially it means you should not use your obtained clusters to explain the variation.
```{r}
# sce <- readRDS(url("https://figshare.com/ndownloader/files/41395260"))
# cellline <- Seurat::as.Seurat(sce)
data(cellline, package = "ClusterDE")
```

# Run the regular Seurat pipeline

We perform the default Seurat clustering. Please note that ClusterDE is designed for 1 vs 1 comparison; therefore, we set the resolution as 0.2 here to obtain two clusters for illustration purpose.
```{r}
RNGkind("L'Ecuyer-CMRG")
set.seed(123)
cellline <- Seurat::UpdateSeuratObject(cellline)
cellline <- Seurat::NormalizeData(cellline)
cellline <- Seurat::FindVariableFeatures(cellline)
cellline <- Seurat::ScaleData(cellline)
cellline <- Seurat::RunPCA(cellline)
cellline <- Seurat::FindNeighbors(cellline)
cellline <- Seurat::FindClusters(cellline, resolution = 0.2)
cellline <- Seurat::RunUMAP(cellline, dims = 1:10)
Seurat::DimPlot(cellline, reduction = "umap") + ggplot2::ggtitle("Clustering result")
```

From the UMAP, the two clusters seem to be dubious. Although we do not expect the existence of cell types, when we perform Seurat DE test between the two clusters, and we get > 1000 genes with FDR < 0.05. It means that the double-dipping introduces a huge number of discoveries.

```{r}
original_markers <- Seurat::FindMarkers(
  cellline,
  ident.1 = 0,
  ident.2 = 1,
  min.pct = 0,
  logfc.threshold = 0
)
message(paste0("Number of DE gene is ", sum(original_markers$p_val_adj < 0.05)))
```

# Find DEGs using ClusterDE

We can use `findMarkers()` to perform null-calibrated post-clustering differential expression. The result table is sorted by contrast scores.
```{r}
res <- ClusterDE::findMarkers(cellline, ident.1 = 0, ident.2 = 1)
```

We can observe from the results that all genes have a frequency record of 0, which means that we do not detect any DE genes.
```{r}
message(paste0("Number of DE gene is ", sum(res$record > 0)))
```

We can also visualize the distribution of contrast scores (diff between the -log p-values from real and null). It is roughly symmetric around 0.
```{r}
ggplot2::ggplot(data = res, ggplot2::aes(x = cs)) +
  ggplot2::geom_histogram(fill = "white", color = "black") +
  ggplot2::theme_bw() +
  ggplot2::ggtitle("Distribution of constrast scores")
```

# Session information
```{r}
sessionInfo()
```
