<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Perform ClusterDE on a cell line dataset â€¢ ClusterDE</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Perform ClusterDE on a cell line dataset">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">ClusterDE</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.4</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/ClusterDE-cellline.html">Perform ClusterDE on a cell line dataset</a></li>
    <li><a class="dropdown-item" href="../articles/ClusterDE-onedomain.html">Perform ClusterDE on a single-domain spatial dataset</a></li>
    <li><a class="dropdown-item" href="../articles/ClusterDE-PBMC.html">Perform ClusterDE on a PBMC dataset</a></li>
    <li><a class="dropdown-item" href="../articles/ClusterDE-twodomains.html">Perform ClusterDE on a two-domain spatial dataset</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/SONGDONGYUAN1994/ClusterDE/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">

<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Perform ClusterDE on a cell line dataset</h1>
                        <h4 data-toc-skip class="author">Dongyuan Song</h4>
            <address class="author_afil">
      Department of Genetics &amp; Genome Sciences, UConn Health<br><a class="author_email" href="mailto:#"></a><a href="mailto:dongyuansong@ucla.edu">dongyuansong@ucla.edu</a>
      </address>
                  
            <h4 data-toc-skip class="date">11 December 2025</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/SONGDONGYUAN1994/ClusterDE/blob/HEAD/vignettes/ClusterDE-cellline.Rmd" class="external-link"><code>vignettes/ClusterDE-cellline.Rmd</code></a></small>
      <div class="d-none name"><code>ClusterDE-cellline.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="download-data">Download data<a class="anchor" aria-label="anchor" href="#download-data"></a>
</h2>
<p>We download the cell line data set H2228. The original data is from <a href="https://github.com/LuyiTian/sc_mixology" class="external-link">Tian et al., Nature Methods 2019</a> as the gold standard for benchmarking the accuracy of clustering. Since the data is from the pure cell line, it should not have cell types, and, of course, between cell type DE genes. <strong>Note</strong>: it does not mean that there are no variations within one cell line; of course there are, e.g., cell cycle, total UMI, etc. However, these variations do not represent <strong>discrete cell groups</strong>, and essentially it means you should not use your obtained clusters to explain the variation.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># sce &lt;- readRDS(url("https://figshare.com/ndownloader/files/41395260"))</span></span>
<span><span class="co"># cellline &lt;- Seurat::as.Seurat(sce)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">cellline</span>, package <span class="op">=</span> <span class="st">"ClusterDE"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="run-the-regular-seurat-pipeline">Run the regular Seurat pipeline<a class="anchor" aria-label="anchor" href="#run-the-regular-seurat-pipeline"></a>
</h2>
<p>We perform the default Seurat clustering. Please note that ClusterDE is designed for 1 vs 1 comparison; therefore, we set the resolution as 0.2 here to obtain two clusters for illustration purpose.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">RNGkind</a></span><span class="op">(</span><span class="st">"L'Ecuyer-CMRG"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span><span class="va">cellline</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/UpdateSeuratObject.html" class="external-link">UpdateSeuratObject</a></span><span class="op">(</span><span class="va">cellline</span><span class="op">)</span></span>
<span><span class="co">#&gt; Validating object structure</span></span>
<span><span class="co">#&gt; Updating object slots</span></span>
<span><span class="co">#&gt; Ensuring keys are in the proper structure</span></span>
<span><span class="co">#&gt; Updating matrix keys for DimReduc 'PCA'</span></span>
<span><span class="co">#&gt; Updating matrix keys for DimReduc 'UMAP'</span></span>
<span><span class="co">#&gt; Ensuring keys are in the proper structure</span></span>
<span><span class="co">#&gt; Ensuring feature names don't have underscores or pipes</span></span>
<span><span class="co">#&gt; Updating slots in originalexp</span></span>
<span><span class="co">#&gt; Updating slots in PCA</span></span>
<span><span class="co">#&gt; Updating slots in UMAP</span></span>
<span><span class="co">#&gt; Setting UMAP DimReduc to global</span></span>
<span><span class="co">#&gt; Validating object structure for Assay 'originalexp'</span></span>
<span><span class="co">#&gt; Validating object structure for DimReduc 'PCA'</span></span>
<span><span class="co">#&gt; Validating object structure for DimReduc 'UMAP'</span></span>
<span><span class="co">#&gt; Object representation is consistent with the most current Seurat version</span></span>
<span><span class="va">cellline</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html" class="external-link">NormalizeData</a></span><span class="op">(</span><span class="va">cellline</span><span class="op">)</span></span>
<span><span class="va">cellline</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html" class="external-link">FindVariableFeatures</a></span><span class="op">(</span><span class="va">cellline</span><span class="op">)</span></span>
<span><span class="va">cellline</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html" class="external-link">ScaleData</a></span><span class="op">(</span><span class="va">cellline</span><span class="op">)</span></span>
<span><span class="co">#&gt; Centering and scaling data matrix</span></span>
<span><span class="va">cellline</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA</a></span><span class="op">(</span><span class="va">cellline</span><span class="op">)</span></span>
<span><span class="co">#&gt; PC_ 1 </span></span>
<span><span class="co">#&gt; Positive:  RPS14, RPL18AP3, RPL36, RPS23, RPL28, LRRC75A-AS1, AC079250.1, FTH1, ZFAS1, EEF2 </span></span>
<span><span class="co">#&gt;     RPL7P9, RPL13A, RPS3AP26, EEF1A1P13, RPS16, RPS23P8, RPL13AP5, RPL29, FTH1P10, RPL13AP25 </span></span>
<span><span class="co">#&gt;     SNHG5, FTH1P8, RPL4, RPS3AP6, AC064799.1, C6orf48, FTH1P7, C1orf56, RPL7AP6, TMSB4X </span></span>
<span><span class="co">#&gt; Negative:  PSMB2, PSMA7, U2AF1, NUDC, RBM8A, CALM1, BUB3, CLIC1, U2AF1L5, XRCC5 </span></span>
<span><span class="co">#&gt;     VPS29, RBM8B, CACYBP, RPA3, SSBP1, PSMC5, MRPL47, PSMD8, BRIX1, CNIH4 </span></span>
<span><span class="co">#&gt;     PCMT1, PSMD13, CYC1, PRDX2, SEPT7, S100A11, VDAC3, PSME2P2, ZWINT, HMGB1 </span></span>
<span><span class="co">#&gt; PC_ 2 </span></span>
<span><span class="co">#&gt; Positive:  NACA, RPL7AP6, SKP1, UBA52, BTF3, SSR2, RPL7A, ARPC3, RPL9P9, PPIA </span></span>
<span><span class="co">#&gt;     PSMD4, EIF1, RPL10, LGALS3BP, RPL10P16, SNRPB2, RPL10P9, S100A11, PPIB, ANXA5 </span></span>
<span><span class="co">#&gt;     EEF2, PSME1, SSBP1, SSR4, RPL7P9, COPE, BSG, MGST1, VPS28, COPS6 </span></span>
<span><span class="co">#&gt; Negative:  SIVA1, HNRNPAB, RPL39L, DEK, CDCA5, TMPO, FAM111A, ASF1B, CENPK, ESCO2 </span></span>
<span><span class="co">#&gt;     BRCA1, H2AFV, RAD51AP1, MT-RNR2, ORC6, CENPX, SNRNP25, FBXO5, RRM1, DIAPH3 </span></span>
<span><span class="co">#&gt;     USP1, CDCA4, TMEM106C, PGP, LSM4, C21orf58, CENPN, BRI3BP, SGO1, CHAF1A </span></span>
<span><span class="co">#&gt; PC_ 3 </span></span>
<span><span class="co">#&gt; Positive:  RPL13AP5, RPL13AP25, AC024293.1, RPL29, RPSAP19, RPS5, RPL18, RPS3AP26, RPS3AP6, RPL15 </span></span>
<span><span class="co">#&gt;     RPL28, DRAP1, RPS11, RPL9P9, RPL13AP7, RPS19, DCBLD2, FXYD5, FEN1, SLBP </span></span>
<span><span class="co">#&gt;     RPS15, COTL1, RPSA, FLNA, RPL7AP6, RPL36, C1orf21, CPA4, ORC6, RPS16 </span></span>
<span><span class="co">#&gt; Negative:  SMIM22, TSPAN13, ST14, PERP, CRB3, MT-CO1, SERINC2, ATP1B1, CDH1, F11R </span></span>
<span><span class="co">#&gt;     B2M, MT-RNR2, SPINT1, NMB, PLA2G16, SPDEF, CD55, ADGRF1, TSPAN1, LIMA1 </span></span>
<span><span class="co">#&gt;     ERBB3, ERO1A, ASS1, CDA, ALCAM, SYNGR2, MT-CO2, CDH3, C3, LSR </span></span>
<span><span class="co">#&gt; PC_ 4 </span></span>
<span><span class="co">#&gt; Positive:  IFNGR1, NAMPT, NAP1L1, CPD, LMAN1, CALR, ITGA2, NAMPTP1, ITM2B, RRM1 </span></span>
<span><span class="co">#&gt;     C3, RHOBTB3, CTHRC1, EEF2, HSD17B11, C1S, IFI16, SMC2, CPE, EPHX1 </span></span>
<span><span class="co">#&gt;     DST, HLA-DMB, NUCB2, MT-ND6, TMEM45A, BRCA1, CDK5RAP2, HINT1, C1R, FAM111A </span></span>
<span><span class="co">#&gt; Negative:  S100A16, TMA7, TIMM8B, PFN1, SLIRP, GPX1, LAMC2, POLR2L, MRPL52, RPS19 </span></span>
<span><span class="co">#&gt;     CDH1, TOMM40, ATP5MD, HSPE1, NAA10, GPX1P1, RPS16, MRPL12, MCRIP2, PDCD5 </span></span>
<span><span class="co">#&gt;     RPL18, PLEC, S100A13, RPL36AL, LAD1, MGLL, BOLA2B, MISP, MRPL36, SEC61G </span></span>
<span><span class="co">#&gt; PC_ 5 </span></span>
<span><span class="co">#&gt; Positive:  HINT1, COX5B, TXN, SOD1, NDUFA4, NDUFS6, ATP5PO, ATP5MC3, S100A10, ATP6V0E1 </span></span>
<span><span class="co">#&gt;     CYB5A, SLIRP, NDUFB4, ATP5MC1, HSPE1, TXNP6, POMP, POLR2L, RPS14, HSPE1P4 </span></span>
<span><span class="co">#&gt;     RPS15, CBR1, NDUFB3, HSPE1P3, ATP5PD, COX7B, ADGRF1, PSMB9, AC079250.1, NDUFAB1 </span></span>
<span><span class="co">#&gt; Negative:  BTG1, PPP1R15A, EIF1, JUN, CEBPG, H3F3B, TMEM132A, C6orf48, HIST2H4B, SGK1 </span></span>
<span><span class="co">#&gt;     KPNA4, PMEPA1, KLF6, CDKN1A, WARS, PEA15, GARS, MAP1LC3B, SNHG12, SERTAD1 </span></span>
<span><span class="co">#&gt;     LAMC2, EPB41L4A-AS1, NAP1L1, SNHG5, KLF10, SLC7A5, KIF5B, ATP2B1, EIF5, ABL2</span></span>
<span><span class="co">#&gt; Warning: Key 'PC_' taken, using 'pca_' instead</span></span>
<span><span class="va">cellline</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html" class="external-link">FindNeighbors</a></span><span class="op">(</span><span class="va">cellline</span><span class="op">)</span></span>
<span><span class="co">#&gt; Computing nearest neighbor graph</span></span>
<span><span class="co">#&gt; Computing SNN</span></span>
<span><span class="va">cellline</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/FindClusters.html" class="external-link">FindClusters</a></span><span class="op">(</span><span class="va">cellline</span>, resolution <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span></span>
<span><span class="co">#&gt; Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Number of nodes: 758</span></span>
<span><span class="co">#&gt; Number of edges: 24895</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Running Louvain algorithm...</span></span>
<span><span class="co">#&gt; Maximum modularity in 10 random starts: 0.8257</span></span>
<span><span class="co">#&gt; Number of communities: 2</span></span>
<span><span class="co">#&gt; Elapsed time: 0 seconds</span></span>
<span><span class="va">cellline</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span><span class="va">cellline</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: The default method for RunUMAP has changed from calling Python UMAP via reticulate to the R-native UWOT using the cosine metric</span></span>
<span><span class="co">#&gt; To use Python UMAP via reticulate, set umap.method to 'umap-learn' and metric to 'correlation'</span></span>
<span><span class="co">#&gt; This message will be shown once per session</span></span>
<span><span class="co">#&gt; 12:55:22 UMAP embedding parameters a = 0.9922 b = 1.112</span></span>
<span><span class="co">#&gt; 12:55:22 Read 758 rows and found 10 numeric columns</span></span>
<span><span class="co">#&gt; 12:55:22 Using Annoy for neighbor search, n_neighbors = 30</span></span>
<span><span class="co">#&gt; 12:55:22 Building Annoy index with metric = cosine, n_trees = 50</span></span>
<span><span class="co">#&gt; 0%   10   20   30   40   50   60   70   80   90   100%</span></span>
<span><span class="co">#&gt; [----|----|----|----|----|----|----|----|----|----|</span></span>
<span><span class="co">#&gt; **************************************************|</span></span>
<span><span class="co">#&gt; 12:55:22 Writing NN index file to temp file /tmp/Rtmp8qoprS/file25396e5a0f8676</span></span>
<span><span class="co">#&gt; 12:55:22 Searching Annoy index using 1 thread, search_k = 3000</span></span>
<span><span class="co">#&gt; 12:55:23 Annoy recall = 100%</span></span>
<span><span class="co">#&gt; 12:55:23 Commencing smooth kNN distance calibration using 1 thread with target n_neighbors = 30</span></span>
<span><span class="co">#&gt; 12:55:24 Initializing from normalized Laplacian + noise (using RSpectra)</span></span>
<span><span class="co">#&gt; 12:55:24 Commencing optimization for 500 epochs, with 27914 positive edges</span></span>
<span><span class="co">#&gt; 12:55:24 Using rng type: pcg</span></span>
<span><span class="co">#&gt; 12:55:25 Optimization finished</span></span>
<span><span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">cellline</span>, reduction <span class="op">=</span> <span class="st">"umap"</span><span class="op">)</span> <span class="op">+</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Clustering result"</span><span class="op">)</span></span></code></pre></div>
<p><img src="ClusterDE-cellline_files/figure-html/unnamed-chunk-3-1.png" class="r-plt" width="700"></p>
<p>From the UMAP, the two clusters seem to be dubious. Although we do not expect the existence of cell types, when we perform Seurat DE test between the two clusters, and we get &gt; 1000 genes with FDR &lt; 0.05. It means that the double-dipping introduces a huge number of discoveries.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">original_markers</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/FindMarkers.html" class="external-link">FindMarkers</a></span><span class="op">(</span></span>
<span>  <span class="va">cellline</span>,</span>
<span>  ident.1 <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  ident.2 <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  min.pct <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  logfc.threshold <span class="op">=</span> <span class="fl">0</span></span>
<span><span class="op">)</span></span>
<span><span class="va">original_markers</span> <span class="op">&lt;-</span> <span class="va">original_markers</span><span class="op">[</span><span class="va">original_markers</span><span class="op">$</span><span class="va">avg_log2FC</span> <span class="op">&gt;</span> <span class="fl">0</span>,<span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">message</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Number of DE gene is "</span>, <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">original_markers</span><span class="op">$</span><span class="va">p_val_adj</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Number of DE gene is 109</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="find-degs-using-clusterde">Find DEGs using ClusterDE<a class="anchor" aria-label="anchor" href="#find-degs-using-clusterde"></a>
</h2>
<p>We can use <code><a href="../reference/findMarkers.html">findMarkers()</a></code> to perform null-calibrated post-clustering differential expression. The result table is sorted by contrast scores.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">ClusterDE</span><span class="fu">::</span><span class="fu"><a href="../reference/findMarkers.html">findMarkers</a></span><span class="op">(</span><span class="va">cellline</span>, ident.1 <span class="op">=</span> <span class="fl">0</span>, ident.2 <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="co">#&gt; 1 genes have no more than 2 non-zero values; ignore fitting and return all 0s.</span></span>
<span><span class="co">#&gt; 98.5% of genes are used in correlation modelling.</span></span>
<span><span class="co">#&gt; 0/1: Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck</span></span>
<span><span class="co">#&gt; 0/1: </span></span>
<span><span class="co">#&gt; 0/1: Number of nodes: 758</span></span>
<span><span class="co">#&gt; 0/1: Number of edges: 29534</span></span>
<span><span class="co">#&gt; 0/1: </span></span>
<span><span class="co">#&gt; 0/1: Running Louvain algorithm...</span></span>
<span><span class="co">#&gt; 0/1: Maximum modularity in 10 random starts: 0.7153</span></span>
<span><span class="co">#&gt; 0/1: Number of communities: 2</span></span>
<span><span class="co">#&gt; 0/1: Elapsed time: 0 seconds</span></span>
<span><span class="co">#&gt; 0/1: Normalizing layer: counts</span></span>
<span><span class="co">#&gt; 0/1: Finding variable features for layer counts</span></span>
<span><span class="co">#&gt; 0/1: Centering and scaling data matrix</span></span>
<span><span class="co">#&gt; 0/1: PC_ 1 </span></span>
<span><span class="co">#&gt; Positive:  PSMB2, CALM1, U2AF1, NUCKS1, RBM8A, CNIH4, PSMA7, MRPL47, VPS29, RPA3 </span></span>
<span><span class="co">#&gt;     XRCC5, U2AF1L5, CLIC1, HMGB1, NDUFA8, PTMAP2, BUB3, SEPT7, RANBP1, EIF5 </span></span>
<span><span class="co">#&gt;     CACYBP, ATP5PO, NDUFAB1, SNRPF, TXN, CYC1, NME1, SSBP1, NUDC, RBM8B </span></span>
<span><span class="co">#&gt; Negative:  RPL34, LRRC75A-AS1, RPL23, AC079250.1, RPS3AP26, ZFAS1, RPL7P9, RPL13A, RPL28, EEF2 </span></span>
<span><span class="co">#&gt;     RPL4, RPS6, RPL10, RPS16, FTH1, SNHG5, RPS3AP6, RPL7AP6, RPL10P16, RPL13AP5 </span></span>
<span><span class="co">#&gt;     RPL13AP25, RPS11, CEBPD, RPS3, RPL9P9, FTH1P10, TMSB4X, RPL10P9, FTH1P8, FTH1P7 </span></span>
<span><span class="co">#&gt; PC_ 2 </span></span>
<span><span class="co">#&gt; Positive:  SSR2, BTF3, RPL7AP6, UBA52, PSMD4, SKP1, RPL10P16, RPL7A, RPL10, RPL9P9 </span></span>
<span><span class="co">#&gt;     EEF2, BSG, RPL10P9, COPE, APEX1, VPS28, ANXA5, EIF1, PPIB, RPL7AP30 </span></span>
<span><span class="co">#&gt;     ARPC3, S100A11, SSBP1, PPIA, SNRPB2, PSMD8, MGST1, RPL7P9, MYDGF, CTSD </span></span>
<span><span class="co">#&gt; Negative:  SIVA1, HNRNPAB, RPL39L, CDCA5, CDCA4, DEK, ASF1B, TMPO, CENPK, FAM111A </span></span>
<span><span class="co">#&gt;     BRCA1, CKLF, H2AFV, CENPX, FBXO5, ESCO2, MND1, DIAPH3, RRM1, USP1 </span></span>
<span><span class="co">#&gt;     MIS18BP1, LSM4, TMEM106C, ORC6, RAD51AP1, MT-CO2, CDCA8, NCAPG, TEX30, ZWINT </span></span>
<span><span class="co">#&gt; PC_ 3 </span></span>
<span><span class="co">#&gt; Positive:  SERINC2, PERP, TSPAN13, ATP1B1, SMIM22, B2M, SPINT1, ST14, F11R, SPTSSA </span></span>
<span><span class="co">#&gt;     ADGRF1, CD55, CRB3, NMB, ERBB3, SLC12A2, CXADR, C3, GPX3, CP </span></span>
<span><span class="co">#&gt;     TAX1BP1, MT-CO2, TMED10, KDM5B, CPD, SDC4, PTBP3, PLA2G16, GOLGB1, MT-CYB </span></span>
<span><span class="co">#&gt; Negative:  RPL14, RPL13AP5, RPL34, RPSAP19, RPL13AP25, SLBP, RPL9P9, RPS3AP26, DRAP1, RPL15 </span></span>
<span><span class="co">#&gt;     AC024293.1, RPS5, RPSA, RPL18, RPL28, RPS3AP6, AC079250.1, SRSF2, PDCD5, RPS16 </span></span>
<span><span class="co">#&gt;     CFL1, RPS11, COTL1, BASP1, CDCA7, NME1, RPS3, FEN1, MCM3, ORC6 </span></span>
<span><span class="co">#&gt; PC_ 4 </span></span>
<span><span class="co">#&gt; Positive:  S100A16, MRPL52, PFN1, TIMM8B, COX6B1, SLIRP, PDCD5, RPS16, LRRC59, CDA </span></span>
<span><span class="co">#&gt;     HSPE1, MGLL, GPX1, RPL18, H3F3B, TOMM40, S100A13, GPX1P1, HSPE1P3, UQCRFS1 </span></span>
<span><span class="co">#&gt;     MRPS23, SNRPG, NDUFB9, MISP, CHCHD10, COX7B, CLTB, TSSC4, AC027309.2, POLR2L </span></span>
<span><span class="co">#&gt; Negative:  IFNGR1, NAMPT, NAMPTP1, CPD, C1R, ITM2B, C1S, RHOBTB3, NAP1L1, CALR </span></span>
<span><span class="co">#&gt;     ITGA2, SMC3, NCOA7, NUCB2, MCM3, CASP4, MT-ND6, ANKRD36B, LMAN1, CP </span></span>
<span><span class="co">#&gt;     RFC4, RRM1, TMEM45A, BRCA1, IFI16, CTHRC1, GLA, ADH5, CDK5RAP2, FANCL </span></span>
<span><span class="co">#&gt; PC_ 5 </span></span>
<span><span class="co">#&gt; Positive:  KPNA4, PPP1R15A, RSL1D1, BTG1, RPL15, TMEM132A, NIPBL, RPSAP19, RPF2, ZYX </span></span>
<span><span class="co">#&gt;     NAP1L1, C6orf48, SNHG7, RAD21, NGDN, HDAC2, PHF20L1, ARHGEF2, MT-ND6, NOP53 </span></span>
<span><span class="co">#&gt;     CALR, CEBPG, CALD1, SLU7, EIF5, SLC2A1, KIF5B, GAR1, SGK1, CDKN1A </span></span>
<span><span class="co">#&gt; Negative:  HINT1, NDUFS6, ATP6V0E1, SOD1, ATP5PO, CYB5A, NDUFB4, ATP5MC3, COX6B1, POLR2L </span></span>
<span><span class="co">#&gt;     FTH1, FTH1P8, POMP, NDUFA4, COX4I1, FTH1P10, CBR1, DDT, REX1BD, CARHSP1 </span></span>
<span><span class="co">#&gt;     PSMB9, COX17, ATP5MC1, CHCHD10, NDUFB3, CLTB, PPIC, ADGRF1, COX7B, SLIRP</span></span>
<span><span class="co">#&gt; 0/1: Computing nearest neighbor graph</span></span>
<span><span class="co">#&gt; 0/1: Computing SNN</span></span></code></pre></div>
<p>We can observe from the results that all genes have a frequency record of 0, which means that we do not detect any DE genes.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">message</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Number of DE gene is "</span>, <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">record</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Number of DE gene is 0</span></span></code></pre></div>
<p>We can also visualize the distribution of contrast scores (diff between the -log p-values from real and null). It is roughly symmetric around 0.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">res</span>, <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">cs</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"white"</span>, color <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Distribution of constrast scores"</span><span class="op">)</span></span>
<span><span class="co">#&gt; `stat_bin()` using `bins = 30`. Pick better value `binwidth`.</span></span></code></pre></div>
<p><img src="ClusterDE-cellline_files/figure-html/unnamed-chunk-7-1.png" class="r-plt" width="700"></p>
</div>
<div class="section level2">
<h2 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.3.1 (2023-06-16)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">#&gt; Running under: Red Hat Enterprise Linux 8.10 (Ootpa)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /sw/pkgs/arc/stacks/gcc/10.3.0/R/4.3.1/lib64/R/lib/libRblas.so </span></span>
<span><span class="co">#&gt; LAPACK: /sw/pkgs/arc/stacks/gcc/10.3.0/R/4.3.1/lib64/R/lib/libRlapack.so;  LAPACK version 3.11.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Random number generation:</span></span>
<span><span class="co">#&gt;  RNG:     L'Ecuyer-CMRG </span></span>
<span><span class="co">#&gt;  Normal:  Inversion </span></span>
<span><span class="co">#&gt;  Sample:  Rejection </span></span>
<span><span class="co">#&gt;  </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">#&gt;  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">#&gt;  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">#&gt; [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: America/Detroit</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] future_1.67.0    BiocStyle_2.30.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] RColorBrewer_1.1-3     jsonlite_2.0.0         magrittr_2.0.4        </span></span>
<span><span class="co">#&gt;   [4] spatstat.utils_3.2-0   farver_2.1.2           rmarkdown_2.30        </span></span>
<span><span class="co">#&gt;   [7] fs_1.6.6               ragg_1.5.0             vctrs_0.6.5           </span></span>
<span><span class="co">#&gt;  [10] ROCR_1.0-11            spatstat.explore_3.5-3 htmltools_0.5.8.1     </span></span>
<span><span class="co">#&gt;  [13] sass_0.4.10            sctransform_0.4.2      parallelly_1.45.1     </span></span>
<span><span class="co">#&gt;  [16] KernSmooth_2.23-21     bslib_0.9.0            htmlwidgets_1.6.4     </span></span>
<span><span class="co">#&gt;  [19] desc_1.4.3             ica_1.0-3              plyr_1.8.9            </span></span>
<span><span class="co">#&gt;  [22] plotly_4.11.0          zoo_1.8-14             cachem_1.1.0          </span></span>
<span><span class="co">#&gt;  [25] igraph_2.2.1           mime_0.13              lifecycle_1.0.4       </span></span>
<span><span class="co">#&gt;  [28] pkgconfig_2.0.3        Matrix_1.6-5           R6_2.6.1              </span></span>
<span><span class="co">#&gt;  [31] fastmap_1.2.0          fitdistrplus_1.2-4     shiny_1.11.1          </span></span>
<span><span class="co">#&gt;  [34] digest_0.6.37          patchwork_1.3.2        Seurat_5.3.1          </span></span>
<span><span class="co">#&gt;  [37] tensor_1.5.1           RSpectra_0.16-2        irlba_2.3.5.1         </span></span>
<span><span class="co">#&gt;  [40] kde1d_1.1.1            textshaping_1.0.4      labeling_0.4.3        </span></span>
<span><span class="co">#&gt;  [43] progressr_0.17.0       spatstat.sparse_3.1-0  coop_0.6-3            </span></span>
<span><span class="co">#&gt;  [46] httr_1.4.7             polyclip_1.10-7        abind_1.4-8           </span></span>
<span><span class="co">#&gt;  [49] compiler_4.3.1         withr_3.0.2            backports_1.5.0       </span></span>
<span><span class="co">#&gt;  [52] S7_0.2.0               fastDummies_1.7.5      MASS_7.3-60           </span></span>
<span><span class="co">#&gt;  [55] tools_4.3.1            lmtest_0.9-40          otel_0.2.0            </span></span>
<span><span class="co">#&gt;  [58] httpuv_1.6.16          future.apply_1.20.0    goftest_1.2-3         </span></span>
<span><span class="co">#&gt;  [61] glue_1.8.0             nlme_3.1-162           promises_1.4.0        </span></span>
<span><span class="co">#&gt;  [64] grid_4.3.1             checkmate_2.3.3        Rtsne_0.17            </span></span>
<span><span class="co">#&gt;  [67] cluster_2.1.4          reshape2_1.4.4         generics_0.1.4        </span></span>
<span><span class="co">#&gt;  [70] gtable_0.3.6           spatstat.data_3.1-9    tidyr_1.3.1           </span></span>
<span><span class="co">#&gt;  [73] data.table_1.17.8      sp_2.2-0               spatstat.geom_3.6-0   </span></span>
<span><span class="co">#&gt;  [76] RcppAnnoy_0.0.22       ggrepel_0.9.6          RANN_2.6.2            </span></span>
<span><span class="co">#&gt;  [79] pillar_1.11.1          stringr_1.5.2          spam_2.11-1           </span></span>
<span><span class="co">#&gt;  [82] RcppHNSW_0.6.0         limma_3.58.1           later_1.4.4           </span></span>
<span><span class="co">#&gt;  [85] splines_4.3.1          dplyr_1.1.4            lattice_0.21-8        </span></span>
<span><span class="co">#&gt;  [88] survival_3.5-5         deldir_2.0-4           tidyselect_1.2.1      </span></span>
<span><span class="co">#&gt;  [91] miniUI_0.1.2           pbapply_1.7-4          knitr_1.50            </span></span>
<span><span class="co">#&gt;  [94] gridExtra_2.3          bookdown_0.45          scattermore_1.2       </span></span>
<span><span class="co">#&gt;  [97] xfun_0.53              statmod_1.5.1          matrixStats_1.5.0     </span></span>
<span><span class="co">#&gt; [100] stringi_1.8.7          ClusterDE_0.99.4       lazyeval_0.2.2        </span></span>
<span><span class="co">#&gt; [103] yaml_2.3.10            evaluate_1.0.5         codetools_0.2-19      </span></span>
<span><span class="co">#&gt; [106] tibble_3.3.0           BiocManager_1.30.26    cli_3.6.5             </span></span>
<span><span class="co">#&gt; [109] uwot_0.2.3             xtable_1.8-4           reticulate_1.44.0     </span></span>
<span><span class="co">#&gt; [112] systemfonts_1.3.1      jquerylib_0.1.4        Rcpp_1.1.0            </span></span>
<span><span class="co">#&gt; [115] globals_0.18.0         spatstat.random_3.4-2  png_0.1-8             </span></span>
<span><span class="co">#&gt; [118] rngWELL_0.10-10        spatstat.univar_3.1-4  parallel_4.3.1        </span></span>
<span><span class="co">#&gt; [121] randtoolbox_2.0.5      assertthat_0.2.1       pkgdown_2.2.0         </span></span>
<span><span class="co">#&gt; [124] ggplot2_4.0.0          presto_1.0.0           mvnfast_0.2.8         </span></span>
<span><span class="co">#&gt; [127] dotCall64_1.2          bettermc_1.2.2.9000    gamlss.dist_6.1-1     </span></span>
<span><span class="co">#&gt; [130] listenv_0.9.1          viridisLite_0.4.2      scales_1.4.0          </span></span>
<span><span class="co">#&gt; [133] ggridges_0.5.7         SeuratObject_5.2.0     purrr_1.1.0           </span></span>
<span><span class="co">#&gt; [136] rlang_1.1.6            rvinecopulib_0.7.3.1.0 cowplot_1.2.0</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Dongyuan Song, Kexin Li, Siqi Chen, Yiwen Yang.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
