<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Perform ClusterDE on a single-domain spatial dataset • ClusterDE</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Perform ClusterDE on a single-domain spatial dataset">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">ClusterDE</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/ClusterDE-cellline.html">Perform ClusterDE on a cell line dataset</a></li>
    <li><a class="dropdown-item" href="../articles/ClusterDE-onedomain.html">Perform ClusterDE on a single-domain spatial dataset</a></li>
    <li><a class="dropdown-item" href="../articles/ClusterDE-PBMC.html">Perform ClusterDE on a PBMC dataset</a></li>
    <li><a class="dropdown-item" href="../articles/ClusterDE-twodomains.html">Perform ClusterDE on a two-domain spatial dataset</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/SONGDONGYUAN1994/ClusterDE/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">

<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Perform ClusterDE on a single-domain spatial dataset</h1>
                        <h4 data-toc-skip class="author">Siqi Chen</h4>
            <address class="author_afil">
      Computer Science, Central South University<br><a class="author_email" href="mailto:#"></a><a href="mailto:siqichen4477@gmail.com">siqichen4477@gmail.com</a>
      </address>
                              <h4 data-toc-skip class="author">Dongyuan Song</h4>
            <address class="author_afil">
      Department of Genetics &amp; Genome Sciences, UConn Health<br><a class="author_email" href="mailto:#"></a><a href="mailto:dongyuansong@ucla.edu">dongyuansong@ucla.edu</a>
      </address>
                  
            <h4 data-toc-skip class="date">11 December 2025</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/SONGDONGYUAN1994/ClusterDE/blob/HEAD/vignettes/ClusterDE-onedomain.Rmd" class="external-link"><code>vignettes/ClusterDE-onedomain.Rmd</code></a></small>
      <div class="d-none name"><code>ClusterDE-onedomain.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="download-data">Download data<a class="anchor" aria-label="anchor" href="#download-data"></a>
</h2>
<p>We selected white matter (WM) of 151673 slice from the LIBD Human Dorsolateral Prefrontal Cortex (DLPFC) dataset, which is downloaded in the spatialLIBD R package.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># # Download the spot-level data</span></span>
<span><span class="co"># spe &lt;- spatialLIBD::fetch_data(type = "spe")</span></span>
<span><span class="co">#</span></span>
<span><span class="co"># # Select the white matter layer in the slice 151673</span></span>
<span><span class="co"># sub_151673 &lt;- spe[, spe$sample_id == "151673"]</span></span>
<span><span class="co"># index &lt;- sub_151673$spatialLIBD == "WM"</span></span>
<span><span class="co"># index[which(is.na(index))] &lt;- "NAN"</span></span>
<span><span class="co"># sub_151673 &lt;- sub_151673[, index == "TRUE"]</span></span>
<span><span class="co"># print(sub_151673)</span></span>
<span><span class="co">#</span></span>
<span><span class="co"># # Delete the genes that express rate less than 20%</span></span>
<span><span class="co"># data &lt;- sub_151673@assays@data$counts</span></span>
<span><span class="co"># zero_expre_rate &lt;- apply(data, 1, function(x) {</span></span>
<span><span class="co">#   zero_true &lt;- x == 0</span></span>
<span><span class="co">#   zero_num &lt;- length(which(zero_true == TRUE)) / dim(data)[2]</span></span>
<span><span class="co">#   return(zero_num)</span></span>
<span><span class="co"># })</span></span>
<span><span class="co"># zero_expre_gene_idx &lt;- which(zero_expre_rate &lt; 0.8)</span></span>
<span><span class="co"># print(paste0("the number of gene with zero expression rate: ", sum(zero_expre_rate == 1), sep = ""))</span></span>
<span><span class="co"># sub_151673 &lt;- sub_151673[zero_expre_gene_idx,]</span></span>
<span><span class="co"># print(paste0("the size of data: ", dim(sub_151673)[1], "*", dim(sub_151673)[2], sep = ""))</span></span>
<span><span class="co">#</span></span>
<span><span class="co"># # Construct the SingleCellExperiment object</span></span>
<span><span class="co"># real_sce &lt;- SingleCellExperiment::SingleCellExperiment(list(counts = sub_151673@assays@data$counts))</span></span>
<span><span class="co"># # Add colData information of singlecellexperiment</span></span>
<span><span class="co"># real_sce$spatial1 &lt;- sub_151673@int_colData@listData$spatialCoords[, 2]</span></span>
<span><span class="co"># real_sce$spatial2 &lt;- sub_151673@int_colData@listData$spatialCoords[, 1]</span></span>
<span><span class="co"># real_sce$cell_type &lt;- sub_151673@colData$spatialLIBD</span></span>
<span><span class="co"># SingleCellExperiment::logcounts(real_sce) &lt;- log1p(SingleCellExperiment::counts(real_sce))</span></span>
<span><span class="co"># dlpfc_onedomain &lt;- Seurat::as.Seurat(real_sce)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">dlpfc_onedomain</span>, package <span class="op">=</span> <span class="st">"ClusterDE"</span><span class="op">)</span></span></code></pre></div>
<p>Visualize the real data on spatial mapping.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Visualize the real spatial domains</span></span>
<span><span class="va">domains</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>Xaxis <span class="op">=</span> <span class="va">dlpfc_onedomain</span><span class="op">$</span><span class="va">spatial1</span>, Yaxis <span class="op">=</span> <span class="va">dlpfc_onedomain</span><span class="op">$</span><span class="va">spatial2</span>, Domain <span class="op">=</span> <span class="va">dlpfc_onedomain</span><span class="op">$</span><span class="va">cell_type</span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: SeuratObject</span></span>
<span><span class="co">#&gt; Loading required package: sp</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'SeuratObject'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect, t</span></span>
<span><span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">domains</span>, <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Xaxis</span>, y <span class="op">=</span> <span class="va">Yaxis</span>, col <span class="op">=</span> <span class="va">Domain</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">1.0</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Manual annotation \n (WM in the slice 151673)"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_equal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span></span>
<span>    plot.title <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">10</span>, hjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span>,</span>
<span>    panel.grid <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>    panel.background <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_rect</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"gray90"</span><span class="op">)</span>,</span>
<span>    panel.border <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_rect</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">"black"</span>, fill <span class="op">=</span> <span class="cn">NA</span>, size <span class="op">=</span> <span class="fl">0.6</span><span class="op">)</span>,</span>
<span>    axis.title.x <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>    axis.title.y <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>    axis.ticks.x <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>    axis.ticks.y <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>    axis.text.x <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>    axis.text.y <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="st">"#5791cc"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: The `size` argument of `element_rect()` is deprecated as of ggplot2 3.4.0.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Please use the `linewidth` argument instead.</span></span>
<span><span class="co">#&gt; <span style="color: #555555;">This warning is displayed once every 8 hours.</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;">Call `lifecycle::last_lifecycle_warnings()` to see where this warning was</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;">generated.</span></span></span></code></pre></div>
<p><img src="ClusterDE-onedomain_files/figure-html/unnamed-chunk-3-1.png" class="r-plt" width="700"></p>
</div>
<div class="section level2">
<h2 id="run-the-bayesspace-seurat-pipeline">Run the BayesSpace + Seurat pipeline<a class="anchor" aria-label="anchor" href="#run-the-bayesspace-seurat-pipeline"></a>
</h2>
<p>Firstly, we employed the BayesSpace for spatial clustering. Please note that ClusterDE is designed for 1 vs 1 comparison; therefore, we obtain two spatial clusters for illustration purpose.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Construct the input of BayesSpace based on real dataset</span></span>
<span><span class="co"># The input of BayesSpace is sce object</span></span>
<span><span class="va">dlpfc_cluster</span> <span class="op">&lt;-</span> <span class="fu">SingleCellExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/SingleCellExperiment.html" class="external-link">SingleCellExperiment</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/AssayData.html" class="external-link">GetAssayData</a></span><span class="op">(</span><span class="va">dlpfc_onedomain</span>, layer <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: replacing previous import 'S4Arrays::makeNindexFromArrayViewport' by</span></span>
<span><span class="co">#&gt; 'DelayedArray::makeNindexFromArrayViewport' when loading 'SummarizedExperiment'</span></span>
<span><span class="co"># Add colData information of SingleCellExperiment</span></span>
<span><span class="va">dlpfc_cluster</span><span class="op">$</span><span class="va">row</span> <span class="op">&lt;-</span> <span class="va">dlpfc_onedomain</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">spatial1</span></span>
<span><span class="va">dlpfc_cluster</span><span class="op">$</span><span class="va">col</span> <span class="op">&lt;-</span> <span class="va">dlpfc_onedomain</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">spatial2</span></span>
<span></span>
<span><span class="co"># Log-normalize the count data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span><span class="va">dlpfc_cluster</span> <span class="op">&lt;-</span> <span class="fu">BayesSpace</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BayesSpace/man/spatialPreprocess.html" class="external-link">spatialPreprocess</a></span><span class="op">(</span><span class="va">dlpfc_cluster</span>, platform <span class="op">=</span> <span class="st">"ST"</span>, n.PCs <span class="op">=</span> <span class="fl">7</span>, log.normalize <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="co"># Clustering with BayesSpace</span></span>
<span><span class="va">dlpfc_cluster</span> <span class="op">&lt;-</span> <span class="fu">BayesSpace</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BayesSpace/man/spatialCluster.html" class="external-link">spatialCluster</a></span><span class="op">(</span></span>
<span>  <span class="va">dlpfc_cluster</span>,</span>
<span>  q <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  platform <span class="op">=</span> <span class="st">"ST"</span>,</span>
<span>  d <span class="op">=</span> <span class="fl">7</span>,</span>
<span>  init.method <span class="op">=</span> <span class="st">"mclust"</span>,</span>
<span>  model <span class="op">=</span> <span class="st">"t"</span>,</span>
<span>  gamma <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  nrep <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>  burn.in <span class="op">=</span> <span class="fl">100</span>,</span>
<span>  save.chain <span class="op">=</span> <span class="cn">T</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Neighbors were identified for 0 out of 673 spots.</span></span>
<span><span class="co">#&gt; Fitting model...</span></span>
<span><span class="co">#&gt; Calculating labels using iterations 100 through 1000.</span></span></code></pre></div>
<p>Visualize the spatial clustering results based on the real data.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Visualize the spatial cluster</span></span>
<span><span class="va">clusters</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>Xaxis <span class="op">=</span> <span class="va">dlpfc_cluster</span><span class="op">$</span><span class="va">row</span>, Yaxis <span class="op">=</span> <span class="va">dlpfc_cluster</span><span class="op">$</span><span class="va">col</span>, Clusters <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">dlpfc_cluster</span><span class="op">$</span><span class="va">spatial.cluster</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">clusters</span>, <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Xaxis</span>, y <span class="op">=</span> <span class="va">Yaxis</span>, col <span class="op">=</span> <span class="va">Clusters</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">1.0</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_equal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Real data \n (spatial clusters detected by BayesSpace)"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span></span>
<span>    plot.title <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">10</span>, hjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span>,</span>
<span>    panel.grid <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>    panel.background <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_rect</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"gray90"</span><span class="op">)</span>,</span>
<span>    panel.border <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_rect</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">"black"</span>, fill <span class="op">=</span> <span class="cn">NA</span>, size <span class="op">=</span> <span class="fl">0.6</span><span class="op">)</span>,</span>
<span>    axis.title.x <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>    axis.title.y <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>    axis.ticks.x <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>    axis.ticks.y <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>    axis.text.x <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>    axis.text.y <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#e87d72"</span>, <span class="st">"#54bcc2"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="ClusterDE-onedomain_files/figure-html/unnamed-chunk-5-1.png" class="r-plt" width="700"></p>
<p>Then, we used the common DE method (Wilcoxon Rank Sum Test) to identify domain marker genes between the two spatial clusters.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Identify domain marker genes in the real dataset based on the BayesSpace clustering result, follow Seurat tutorial</span></span>
<span><span class="va">dlpfc_onedomain</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/as.Seurat.html" class="external-link">as.Seurat</a></span><span class="op">(</span><span class="va">dlpfc_cluster</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Keys should be one or more alphanumeric characters followed by an</span></span>
<span><span class="co">#&gt; underscore, setting key from PC to PC_</span></span>
<span><span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/Idents.html" class="external-link">Idents</a></span><span class="op">(</span><span class="va">dlpfc_onedomain</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"spatial.cluster"</span></span>
<span><span class="va">original_markers</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/FindMarkers.html" class="external-link">FindMarkers</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">dlpfc_onedomain</span>,</span>
<span>  ident.1 <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  ident.2 <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  test.use <span class="op">=</span> <span class="st">"wilcox"</span>,</span>
<span>  logfc.threshold <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  min.pct <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  min.cells.feature <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  min.cells.group <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="op">)</span></span>
<span><span class="va">original_markers</span> <span class="op">&lt;-</span> <span class="va">original_markers</span><span class="op">[</span><span class="va">original_markers</span><span class="op">$</span><span class="va">avg_log2FC</span> <span class="op">&gt;</span> <span class="fl">0</span>,<span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">message</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Number of DE gene is "</span>, <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">original_markers</span><span class="op">$</span><span class="va">p_val_adj</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Number of DE gene is 21</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="find-degs-using-clusterde">Find DEGs using ClusterDE<a class="anchor" aria-label="anchor" href="#find-degs-using-clusterde"></a>
</h2>
<p>We can use <code><a href="../reference/findMarkers.html">findMarkers()</a></code> from ClusterDE with <code>spatial</code> specified as name of X and Y coordinates in the data meta columns. Here in the example, coordinates are stored in <code>row</code> and <code>col</code>.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">ClusterDE</span><span class="fu">::</span><span class="fu"><a href="../reference/findMarkers.html">findMarkers</a></span><span class="op">(</span><span class="va">dlpfc_onedomain</span>, ident.1 <span class="op">=</span> <span class="fl">1</span>, ident.2 <span class="op">=</span> <span class="fl">2</span>, spatial <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"row"</span>, <span class="st">"col"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; 100% of genes are used in correlation modelling.</span></span>
<span><span class="co">#&gt; 0/1: Neighbors were identified for 0 out of 673 spots.</span></span>
<span><span class="co">#&gt; 0/1: Fitting model...</span></span>
<span><span class="co">#&gt; 0/1: Calculating labels using iterations 100 through 1000.</span></span>
<span><span class="co">#&gt; 0/1: Normalizing layer: counts</span></span></code></pre></div>
<p>We can observe from the results that the cell compositions are nearly identical, leaving only one gene with detectable differential expression.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">message</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Number of DE gene is "</span>, <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">record</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Number of DE gene is 6</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.3.1 (2023-06-16)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">#&gt; Running under: Red Hat Enterprise Linux 8.10 (Ootpa)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /sw/pkgs/arc/stacks/gcc/10.3.0/R/4.3.1/lib64/R/lib/libRblas.so </span></span>
<span><span class="co">#&gt; LAPACK: /sw/pkgs/arc/stacks/gcc/10.3.0/R/4.3.1/lib64/R/lib/libRlapack.so;  LAPACK version 3.11.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">#&gt;  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">#&gt;  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">#&gt; [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: America/Detroit</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] SeuratObject_5.2.0 sp_2.2-0           BiocStyle_2.30.0  </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] fs_1.6.6                    matrixStats_1.5.0          </span></span>
<span><span class="co">#&gt;   [3] spatstat.sparse_3.1-0       bitops_1.0-9               </span></span>
<span><span class="co">#&gt;   [5] httr_1.4.7                  RColorBrewer_1.1-3         </span></span>
<span><span class="co">#&gt;   [7] backports_1.5.0             tools_4.3.1                </span></span>
<span><span class="co">#&gt;   [9] sctransform_0.4.2           R6_2.6.1                   </span></span>
<span><span class="co">#&gt;  [11] DirichletReg_0.7-2          lazyeval_0.2.2             </span></span>
<span><span class="co">#&gt;  [13] uwot_0.2.3                  rhdf5filters_1.14.1        </span></span>
<span><span class="co">#&gt;  [15] withr_3.0.2                 ClusterDE_0.99.3           </span></span>
<span><span class="co">#&gt;  [17] gridExtra_2.3               progressr_0.17.0           </span></span>
<span><span class="co">#&gt;  [19] cli_3.6.5                   Biobase_2.62.0             </span></span>
<span><span class="co">#&gt;  [21] textshaping_1.0.4           spatstat.explore_3.5-3     </span></span>
<span><span class="co">#&gt;  [23] fastDummies_1.7.5           sandwich_3.1-1             </span></span>
<span><span class="co">#&gt;  [25] labeling_0.4.3              sass_0.4.10                </span></span>
<span><span class="co">#&gt;  [27] Seurat_5.3.1                S7_0.2.0                   </span></span>
<span><span class="co">#&gt;  [29] spatstat.data_3.1-9         ggridges_0.5.7             </span></span>
<span><span class="co">#&gt;  [31] pbapply_1.7-4               pkgdown_2.2.0              </span></span>
<span><span class="co">#&gt;  [33] systemfonts_1.3.1           scater_1.30.1              </span></span>
<span><span class="co">#&gt;  [35] parallelly_1.45.1           limma_3.58.1               </span></span>
<span><span class="co">#&gt;  [37] RSQLite_2.4.5               generics_0.1.4             </span></span>
<span><span class="co">#&gt;  [39] ica_1.0-3                   spatstat.random_3.4-2      </span></span>
<span><span class="co">#&gt;  [41] dplyr_1.1.4                 Matrix_1.6-5               </span></span>
<span><span class="co">#&gt;  [43] ggbeeswarm_0.7.3            S4Vectors_0.40.2           </span></span>
<span><span class="co">#&gt;  [45] abind_1.4-8                 lifecycle_1.0.4            </span></span>
<span><span class="co">#&gt;  [47] yaml_2.3.10                 edgeR_4.0.16               </span></span>
<span><span class="co">#&gt;  [49] SummarizedExperiment_1.32.0 rhdf5_2.46.1               </span></span>
<span><span class="co">#&gt;  [51] SparseArray_1.2.4           BiocFileCache_2.10.2       </span></span>
<span><span class="co">#&gt;  [53] Rtsne_0.17                  grid_4.3.1                 </span></span>
<span><span class="co">#&gt;  [55] blob_1.2.4                  promises_1.4.0             </span></span>
<span><span class="co">#&gt;  [57] dqrng_0.4.1                 crayon_1.5.3               </span></span>
<span><span class="co">#&gt;  [59] miniUI_0.1.2                lattice_0.21-8             </span></span>
<span><span class="co">#&gt;  [61] beachmat_2.18.1             cowplot_1.2.0              </span></span>
<span><span class="co">#&gt;  [63] kde1d_1.1.1                 pillar_1.11.1              </span></span>
<span><span class="co">#&gt;  [65] knitr_1.50                  metapod_1.10.1             </span></span>
<span><span class="co">#&gt;  [67] GenomicRanges_1.54.1        randtoolbox_2.0.5          </span></span>
<span><span class="co">#&gt;  [69] xgboost_3.1.2.1             future.apply_1.20.0        </span></span>
<span><span class="co">#&gt;  [71] codetools_0.2-19            glue_1.8.0                 </span></span>
<span><span class="co">#&gt;  [73] spatstat.univar_3.1-4       data.table_1.17.8          </span></span>
<span><span class="co">#&gt;  [75] vctrs_0.6.5                 png_0.1-8                  </span></span>
<span><span class="co">#&gt;  [77] spam_2.11-1                 gtable_0.3.6               </span></span>
<span><span class="co">#&gt;  [79] assertthat_0.2.1            cachem_1.1.0               </span></span>
<span><span class="co">#&gt;  [81] xfun_0.53                   S4Arrays_1.2.1             </span></span>
<span><span class="co">#&gt;  [83] mime_0.13                   coop_0.6-3                 </span></span>
<span><span class="co">#&gt;  [85] rngWELL_0.10-10             coda_0.19-4.1              </span></span>
<span><span class="co">#&gt;  [87] survival_3.5-5              SingleCellExperiment_1.24.0</span></span>
<span><span class="co">#&gt;  [89] maxLik_1.5-2.1              statmod_1.5.1              </span></span>
<span><span class="co">#&gt;  [91] bluster_1.12.0              BayesSpace_1.12.0          </span></span>
<span><span class="co">#&gt;  [93] fitdistrplus_1.2-4          ROCR_1.0-11                </span></span>
<span><span class="co">#&gt;  [95] bettermc_1.2.2.9000         nlme_3.1-162               </span></span>
<span><span class="co">#&gt;  [97] bit64_4.6.0-1               filelock_1.0.3             </span></span>
<span><span class="co">#&gt;  [99] RcppAnnoy_0.0.22            GenomeInfoDb_1.38.8        </span></span>
<span><span class="co">#&gt; [101] bslib_0.9.0                 irlba_2.3.5.1              </span></span>
<span><span class="co">#&gt; [103] vipor_0.4.7                 KernSmooth_2.23-21         </span></span>
<span><span class="co">#&gt; [105] otel_0.2.0                  BiocGenerics_0.48.1        </span></span>
<span><span class="co">#&gt; [107] DBI_1.2.3                   tidyselect_1.2.1           </span></span>
<span><span class="co">#&gt; [109] bit_4.6.0                   compiler_4.3.1             </span></span>
<span><span class="co">#&gt; [111] curl_7.0.0                  BiocNeighbors_1.20.2       </span></span>
<span><span class="co">#&gt; [113] desc_1.4.3                  DelayedArray_0.28.0        </span></span>
<span><span class="co">#&gt; [115] plotly_4.11.0               bookdown_0.45              </span></span>
<span><span class="co">#&gt; [117] checkmate_2.3.3             scales_1.4.0               </span></span>
<span><span class="co">#&gt; [119] lmtest_0.9-40               mvnfast_0.2.8              </span></span>
<span><span class="co">#&gt; [121] rvinecopulib_0.7.3.1.0      stringr_1.5.2              </span></span>
<span><span class="co">#&gt; [123] digest_0.6.37               goftest_1.2-3              </span></span>
<span><span class="co">#&gt; [125] presto_1.0.0                spatstat.utils_3.2-0       </span></span>
<span><span class="co">#&gt; [127] rmarkdown_2.30              XVector_0.42.0             </span></span>
<span><span class="co">#&gt; [129] htmltools_0.5.8.1           pkgconfig_2.0.3            </span></span>
<span><span class="co">#&gt; [131] sparseMatrixStats_1.14.0    MatrixGenerics_1.14.0      </span></span>
<span><span class="co">#&gt; [133] dbplyr_2.5.1                fastmap_1.2.0              </span></span>
<span><span class="co">#&gt; [135] rlang_1.1.6                 htmlwidgets_1.6.4          </span></span>
<span><span class="co">#&gt; [137] shiny_1.11.1                DelayedMatrixStats_1.24.0  </span></span>
<span><span class="co">#&gt; [139] farver_2.1.2                jquerylib_0.1.4            </span></span>
<span><span class="co">#&gt; [141] zoo_1.8-14                  jsonlite_2.0.0             </span></span>
<span><span class="co">#&gt; [143] BiocParallel_1.36.0         mclust_6.1.1               </span></span>
<span><span class="co">#&gt; [145] BiocSingular_1.18.0         RCurl_1.98-1.17            </span></span>
<span><span class="co">#&gt; [147] magrittr_2.0.4              Formula_1.2-5              </span></span>
<span><span class="co">#&gt; [149] scuttle_1.12.0              GenomeInfoDbData_1.2.11    </span></span>
<span><span class="co">#&gt; [151] dotCall64_1.2               patchwork_1.3.2            </span></span>
<span><span class="co">#&gt; [153] Rhdf5lib_1.24.2             Rcpp_1.1.0                 </span></span>
<span><span class="co">#&gt; [155] viridis_0.6.5               reticulate_1.44.0          </span></span>
<span><span class="co">#&gt; [157] stringi_1.8.7               zlibbioc_1.48.2            </span></span>
<span><span class="co">#&gt; [159] MASS_7.3-60                 gamlss.dist_6.1-1          </span></span>
<span><span class="co">#&gt; [161] plyr_1.8.9                  parallel_4.3.1             </span></span>
<span><span class="co">#&gt; [163] listenv_0.9.1               ggrepel_0.9.6              </span></span>
<span><span class="co">#&gt; [165] deldir_2.0-4                splines_4.3.1              </span></span>
<span><span class="co">#&gt; [167] tensor_1.5.1                locfit_1.5-9.12            </span></span>
<span><span class="co">#&gt; [169] igraph_2.2.1                spatstat.geom_3.6-0        </span></span>
<span><span class="co">#&gt; [171] RcppHNSW_0.6.0              reshape2_1.4.4             </span></span>
<span><span class="co">#&gt; [173] stats4_4.3.1                ScaledMatrix_1.10.0        </span></span>
<span><span class="co">#&gt; [175] evaluate_1.0.5              scran_1.30.2               </span></span>
<span><span class="co">#&gt; [177] BiocManager_1.30.26         httpuv_1.6.16              </span></span>
<span><span class="co">#&gt; [179] miscTools_0.6-28            RANN_2.6.2                 </span></span>
<span><span class="co">#&gt; [181] tidyr_1.3.1                 purrr_1.1.0                </span></span>
<span><span class="co">#&gt; [183] polyclip_1.10-7             future_1.67.0              </span></span>
<span><span class="co">#&gt; [185] scattermore_1.2             ggplot2_4.0.0              </span></span>
<span><span class="co">#&gt; [187] rsvd_1.0.5                  xtable_1.8-4               </span></span>
<span><span class="co">#&gt; [189] RSpectra_0.16-2             later_1.4.4                </span></span>
<span><span class="co">#&gt; [191] viridisLite_0.4.2           ragg_1.5.0                 </span></span>
<span><span class="co">#&gt; [193] tibble_3.3.0                memoise_2.0.1              </span></span>
<span><span class="co">#&gt; [195] beeswarm_0.4.0              IRanges_2.36.0             </span></span>
<span><span class="co">#&gt; [197] cluster_2.1.4               globals_0.18.0</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Dongyuan Song, Kexin Li, Siqi Chen.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
